src/Data/List/Sort/InsertionSort.agda

```diff
@@ -0,0 +1,230 @@
+
+insert-swap : ∀ x y xs → insert x (insert y xs) ≋ insert y (insert x xs)
+insert-swap x y [] with x ≤? y | y ≤? x
+... | yes x≤y | yes y≤x = eq ∷ (Eq.sym eq) ∷ []
+  where eq = antisym x≤y y≤x
+... | no x≤y | yes y≤x = reflₚₜ
+... | yes x≤y | no y≤x = reflₚₜ
+... | no x≤y | no y≤x = (Eq.sym eq) ∷ (eq ∷ [])
+  where eq = ≥-antisym (≰⇒≥ x≤y) (≰⇒≥ y≤x)
+insert-swap x y (z ∷ xs) with x ≤? z in xz | y ≤? z in yz |
+  x ≤? y in xy | y ≤? x in yx
+insert-swap x y (z ∷ xs) | yes x≤z | yes y≤z | yes x≤y | yes y≤x
+  rewrite xy | yx = eq ∷ (Eq.sym eq) ∷ reflₚₜ
+  where eq = antisym x≤y y≤x
+insert-swap x y (z ∷ xs) | yes x≤z | yes y≤z | yes x≤y | no y≤x
+  rewrite xy | yx | yz = reflₚₜ
+insert-swap x y (z ∷ xs) | yes x≤z | yes y≤z | no x≤y | yes y≤x
+  rewrite xy | yx | xz = reflₚₜ
+insert-swap x y (z ∷ xs) | yes x≤z | yes y≤z | no x≤y | no y≤x
+  rewrite xy | yx | yz | xz = Eq.sym eq ∷ eq ∷ (reflₚₜ)
+  where eq = ≥-antisym (≰⇒≥ x≤y) (≰⇒≥ y≤x)  
+insert-swap x y (z ∷ xs) | yes x≤z | no y≤z | yes x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = ⊥-elim (y≤z (transA y≤x x≤z))
+insert-swap x y (z ∷ xs) | yes x≤z | no y≤z | yes x≤y | no y≤x
+  rewrite xy | yx | yz | xz = reflₚₜ
+insert-swap x y (z ∷ xs) | yes x≤z | no y≤z | no x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = ⊥-elim (y≤z (transA y≤x x≤z))
+insert-swap x y (z ∷ xs) | yes x≤z | no y≤z | no x≤y | no y≤x
+  rewrite xy | yx | yz | xz = reflₚₜ
+insert-swap x y (z ∷ xs) | no x≤z | yes y≤z | yes x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = ⊥-elim (x≤z (transA x≤y y≤z))
+insert-swap x y (z ∷ xs) | no x≤z | yes y≤z | yes x≤y | no y≤x
+  rewrite xy | yx | yz | xz = ⊥-elim (x≤z (transA x≤y y≤z))
+insert-swap x y (z ∷ xs) | no x≤z | yes y≤z | no x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = reflₚₜ
+insert-swap x y (z ∷ xs) | no x≤z | yes y≤z | no x≤y | no y≤x
+  rewrite xy | yx | yz | xz = reflₚₜ
+insert-swap x y (z ∷ xs) | no x≤z | no y≤z | yes x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = Eq.refl ∷ (insert-swap x y xs)
+insert-swap x y (z ∷ xs) | no x≤z | no y≤z | yes x≤y | no y≤x
+  rewrite xy | yx | yz | xz = Eq.refl ∷ (insert-swap x y xs)
+insert-swap x y (z ∷ xs) | no x≤z | no y≤z | no x≤y | yes y≤x
+  rewrite xy | yx | yz | xz = Eq.refl ∷ (insert-swap x y xs)
+insert-swap x y (z ∷ xs) | no x≤z | no y≤z | no x≤y | no y≤x
+  rewrite xy | yx | yz | xz = Eq.refl ∷ (insert-swap x y xs)
+
+insert-swap-cong : ∀ {x y x′ y′ xs ys} →
+                   x ≈ x′ → y ≈ y′ → xs ≋ ys →
+                   insert x (insert y xs) ≋ insert y′ (insert x′ ys)
+insert-swap-cong {x} {y} {x′} {y′} {xs} {ys} eq1 eq2 eq3 = begin
+  insert x (insert y xs)   ≈⟨ insert-cong eq1 (insert-cong eq2 eq3) ⟩
+  insert x′ (insert y′ ys) ≈⟨ insert-swap x′ y′ ys ⟩
+  insert y′ (insert x′ ys) ∎
+  where open SetoidReasoning
+
+-- Ideally, we want:
+
+--   property1 : ∀ {xs ys} → xs ↭ ys →
+--               Sorted xs → Sorted ys → xs ≋ ys
+
+-- But the induction over xs ↭ ys is hard to do. So instead we have a
```

Comment:
In what way hard, especially in view of `insert-swap` above? Would the abstract proof via `Sorted` be any worse?

Reply:
Ah... transitivity. A more serious bug in the definition of `Permutation`: transitivity is eliminable!

Reply:
See #2311